Source/Core/Core/IOS/ES/Views.cpp

```diff
@@ -135,16 +135,56 @@ ReturnCode ESDevice::GetV0TicketFromView(const u8* ticket_view, u8* ticket) cons
   return IPC_SUCCESS;
 }
 
+ReturnCode ESDevice::GetV1TicketFromView(const u8* ticket_view, u8* ticket,
+                                         std::optional<u32*> ticket_size) const
+{
+  const u64 title_id = Common::swap64(&ticket_view[offsetof(ES::TicketView, title_id)]);
+  const u64 ticket_id = Common::swap64(&ticket_view[offsetof(ES::TicketView, ticket_id)]);
+
+  const auto installed_ticket = FindSignedTicket(title_id);
+  if (!installed_ticket.IsValid())
+    return ES_NO_TICKET;
+
+  if (ticket_size)
+  {
+    **ticket_size = installed_ticket.GetTicketSize();
+    return IPC_SUCCESS;
+  }
+
+  const std::vector<u8> ticket_bytes = installed_ticket.GetRawTicket(ticket_id);
+  if (ticket_bytes.empty())
+    return ES_NO_TICKET;
+
+  if (!m_title_context.active)
+    return ES_EINVAL;
+
+  // Check for permission to export the ticket.
+  const u32 title_identifier = static_cast<u32>(installed_ticket.GetTitleId());
+  const u32 permitted_title_mask =
+      Common::swap32(ticket_bytes.data() + offsetof(ES::Ticket, permitted_title_mask));
+  const u32 permitted_title_id =
+      Common::swap32(ticket_bytes.data() + offsetof(ES::Ticket, permitted_title_id));
+  const u8 title_export_allowed = ticket_bytes[offsetof(ES::Ticket, title_export_allowed)];
+
+  if (!title_identifier || (title_identifier & ~permitted_title_mask) != permitted_title_id ||
+      (title_export_allowed & 0xF) != 5)
+  {
+    return ES_EACCES;
+  }
+
+  std::copy(ticket_bytes.begin(), ticket_bytes.end(), ticket);
+  return IPC_SUCCESS;
+}
+
 ReturnCode ESDevice::GetTicketFromView(const u8* ticket_view, u8* ticket, u32* ticket_size) const
 {
   const u8 version = ticket_view[offsetof(ES::TicketView, version)];
   if (version == 1)
   {
-    // Currently, we have no support for v1 tickets at all (unlike IOS), so we fake it
-    // and return that there is no ticket.
-    // TODO: implement GetV1TicketFromView when we gain v1 ticket support.
-    ERROR_LOG_FMT(IOS_ES, "GetV1TicketFromView: Unimplemented -- returning -1028");
-    return ES_NO_TICKET;
+    if (ticket == nullptr)
+      return GetV1TicketFromView(ticket_view, ticket, ticket_size);
```

Comment:
This is just a suggestion but I wonder if the code would be clearer as two functions:

```
ReturnCode GetV1TicketFromViewAndExistingTicket(const u8* ticket_view, u8* ticket) const
{
  // same as current function except remove the code for ticket_size
}


ReturnCode GetV1TicketFromViewAndSize(const u8* ticket_view, u32* ticket_size) const
{
  const u64 title_id = Common::swap64(&ticket_view[offsetof(ES::TicketView, title_id)]);

  const auto installed_ticket = FindSignedTicket(title_id);
  if (!installed_ticket.IsValid())
    return ES_NO_TICKET;

   *ticket_size = installed_ticket.GetTicketSize();
    return IPC_SUCCESS;
}
```

Reply:
Another way would be to make `GetTicketFromView`, `GetV0TicketFromView` and `GetV1TicketFromView` a single function.

AFAICT, `GetTicketFromView` set `*ticket_size` just like `GetV1TicketFromView` and `GetV1TicketFromView` behaves like `GetV0TicketFromView` but adds a size check.

Reply:
This does sound like a better way to do it, will change.